### PR TITLE
Add github workflow to verify Espressif ESP32 sample

### PR DESCRIPTION
**Pull request by ewertons**
## Purpose
Add a workflow to validate PRs do not break the Espressif ESP32 sample.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[x] Other... Please describe: CI pipeline improvement.
```

## How to Test
Just create a PR and verify the output of "ESPRESSIF ESP32 CI Tests"

## What to Check
The check has completed successfully.
Additionally verify the end of the check log contains information for flashing a device:

```
Project build complete. To flash, run this command:
/opt/esp/python_env/idf4.4_py3.8_env/bin/python opt/esp/idf/components/esptool_py/esptool/esptool.py -p (PORT) -b 460800 --before default_reset --after hard_reset --chip esp32  write_flash --flash_mode dio --flash_size detect --flash_freq 40m 0x1000 iot-middleware-freertos-samples/demos/projects/ESPRESSIF/esp32/build/bootloader/bootloader.bin 0x8000 iot-middleware-freertos-samples/demos/projects/ESPRESSIF/esp32/build/partition_table/partition-table.bin 0x10000 iot-middleware-freertos-samples/demos/projects/ESPRESSIF/esp32/build/azure_iot_freertos_esp32.bin
or run 'idf.py -p (PORT) flash'
```

### CODE REVIEW

**Review by danewalton — APPROVED** [review]

File: .github/workflows/ci_tests_esp32.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    steps:
+    - uses: docker://espressif/idf:latest
+    - uses: actions/checkout@v2
+    - uses: snickerbockers/submodules-init@v4
+    - name: Build Esp32 Sample
+      run: docker run -v "${PWD}:/iot-middleware-freertos-samples" espressif/idf idf.py build -C /iot-middleware-freertos-samples/demos/projects/ESPRESSIF/esp32
```
Comment (hihigupt) [inline .github/workflows/ci_tests_esp32.yml:14]: is it possible to add this build to following script?
https://github.com/Azure-Samples/iot-middleware-freertos-samples/blob/main/.github/scripts/ci_tests.sh

this way locally we can run this locally using this script.

Additionally, will docker command fail if the script passed to it fails?

File: .github/workflows/ci_tests_esp32.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+name: ESPRESSIF ESP32 CI Tests
+
+on: [pull_request]
+
+jobs:
+  build:
+
+    runs-on: ubuntu-latest
+    steps:
+    - uses: docker://espressif/idf:latest
```
Comment (hihigupt) [inline .github/workflows/ci_tests_esp32.yml:10]: Running on latest is little aggressive, and can potentially break our build system. Can we control this by using some release/stable version. Version which we offically support. I saw they have  image for releases like :
espressif/idf:release-v4.0

Comment (ewertons) [reply .github/workflows/ci_tests_esp32.yml:10]: As per our discussion, fine we will use a fixed version:v4.3

File: .github/workflows/ci_tests_linux.yml [hunk 3]
```diff
@@ -16,3 +17,5 @@ jobs:
       run: .github/scripts/install_software.sh
     - name: Check-in Tests
       run: sudo bash -c ".github/scripts/ci_tests.sh"
+    - name: Build Esp32 Sample
+      run: docker run -v "${PWD}:/iot-middleware-freertos-samples" -w /iot-middleware-freertos-samples espressif/idf .github/scripts/ci_tests.sh -esp
```
Comment (hihigupt) [inline .github/workflows/ci_tests_linux.yml:21]: >/iot-middleware-freertos-samples espressif/idf [](http://example.com/codeflow?start=70&length=46)

working directory should be same to your mounting directory? or I am missing something here?

docker run -v "${PWD}:/iot-middleware-freertos-samples" -w /iot-middleware-freertos-samples .github/scripts/ci_tests.sh -esp

**Comment by hihigupt** [comment]
> set -o errexit # Exit if command failed.

Should we add -all to build all and default should be all. Then we have following documentation.

eg:
# ci_tests.sh [-st] [-nxp] [-pc] [-esp]
#
# if none specified all vendors are build

#! /bin/bash

# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for full license information.
#
# ci_tests.sh [-st] [-nxp] [-pc] [-esp]
#
# if none specified all vendors are build
#
set -o errexit # Exit if command failed.
set -o nounset # Exit if variable not set.
set -o pipefail # Exit if pipe failed.

TEST_SCRIPT_DIR=`dirname "$0"`
TEST_FREERTOS_SRC=`pwd`/libs/FreeRTOS
RUN_BOARDS_BUILD=${@:-"-st -nxp -pc -esp"}


function exit_if_binary_does_not_exist()
{
    local search_dir=$1
    local filename=$2

    if [ $(find "$search_dir" -name "$filename") ]; then 
      echo "$filename is found in $search_dir"
    else
      echo "$filename not found"
      exit 1
    fi
}

function sample_build() {
    local vendor=$1
    local board=$2
    local outdir=$3

    if [ $vendor == "ESPRESSIF" ]
    then
      idf.py build -C ./demos/projects/ESPRESSIF/esp32
    else
      cmake -G Ninja -DBOARD=$board -DVENDOR=$vendor -B$outdir -DFREERTOS_PATH=$TEST_FREERTOS_SRC .
      cmake --build $outdir
    fi
}

for arg in $RUN_BOARDS_BUILD
do
        case "$arg" in
                "-esp")
                            echo -e "::group::Building sample for ESPRESSIF ESP32 port"
    sample_build "ESPRESSIF" "esp32" "build"
    exit_if_binary_does_not_exist "./demos/projects/ESPRESSIF/esp32/build" "azure_iot_freertos_esp32.bin"
                        ;;
                "-nxp")
                    echo -e "::group::Building sample for NXP mimxrt1060 port"
    sample_build "NXP" "mimxrt1060" "build_nxp_mimxrt1060"
    exit_if_binary_does_not_exist "build_nxp_mimxrt1060" "iot-middleware-sample.elf"
    exit_if_binary_does_not_exist "build_nxp_mimxrt1060" "iot-middleware-sample-pnp.elf"
                        ;;
                "-st")
                sample_build "ST" "b-l475e-iot01a" "build_st_b-l475e-iot01a"
                exit_if_binary_does_not_exist "build_st_b-l475e-iot01a" "iot-middleware-sample.elf"
                exit_if_binary_does_not_exist "build_st_b-l475e-iot01a" "iot-middleware-sample-pnp.elf"
                exit_if_binary_does_not_exist "build_st_b-l475e-iot01a" "iot-middleware-sample-gsg.elf"

                    sample_build "ST" "stm32h745i-disco" "build_st_stm32h745i-disco"
    exit_if_binary_does_not_exist "build_st_stm32h745i-disco" "iot-middleware-sample.elf"
    exit_if_binary_does_not_exist "build_st_stm32h745i-disco" "iot-middleware-sample-pnp.elf"

    echo -e "::group::Building sample for ST b-l4s5i-iot01a port"
    sample_build "ST" "b-l4s5i-iot01a" "build_st_b-l4s5i-iot01a"
    exit_if_binary_does_not_exist "build_st_b-l4s5i-iot01a" "iot-middleware-sample.elf"
    exit_if_binary_does_not_exist "build_st_b-l4s5i-iot01a" "iot-middleware-sample-pnp.elf"

                ;;
                "-pc")
                sample_build "PC" "linux" "build_pc_linux"
                exit_if_binary_does_not_exist "build_pc_linux" "iot-middleware-sample"
                exit_if_binary_does_not_exist "build_pc_linux" "iot-middleware-sample-pnp"
                        ;;
        * )
                echo "build for $arg not found";;
        esac
done





---
Refers to: .github/scripts/ci_tests.sh:6 in 0f338b6. [](commit_id = 0f338b6608e35392b5466f2fb40a4aa8be48dc00, deletion_comment = False)

File: .github/workflows/ci_tests_linux.yml (hunk 3 above)
Comment (ewertons) [reply .github/workflows/ci_tests_linux.yml:21]: I'm not sure what you are proposing. 
Your command is missing the image name. The one that is in the code is correct.

File: .github/scripts/ci_tests.sh [hunk 4]
```diff
@@ -2,14 +2,19 @@
...
+# if none specified all vendors are build
 
 set -o errexit # Exit if command failed.
 set -o nounset # Exit if variable not set.
 set -o pipefail # Exit if pipe failed.
 
 TEST_SCRIPT_DIR=`dirname "$0"`
 TEST_FREERTOS_SRC=`pwd`/libs/FreeRTOS
-$TEST_SCRIPT_DIR/fetch_freertos.sh $TEST_FREERTOS_SRC
+RUN_BOARDS_BUILD=${@:-"-st -nxp -pc -esp"}
```
Comment (hihigupt) [inline .github/scripts/ci_tests.sh:16]: $TEST_SCRIPT_DIR/fetch_freertos.sh $TEST_FREERTOS_SRC

this should not be removed.

File: .github/scripts/ci_tests.sh [hunk 5]
```diff
@@ -29,38 +34,69 @@ function sample_build() {
...
+    then
+      idf.py build -C ./demos/projects/ESPRESSIF/esp32
+    else
+      cmake -G Ninja -DBOARD=$board -DVENDOR=$vendor -B$outdir -DFREERTOS_PATH=$TEST_FREERTOS_SRC .
+      cmake --build $outdir
+    fi
 }
 
-######################## Simulations #######################
+function fetch_freertos()
```
Comment (hihigupt) [inline .github/scripts/ci_tests.sh:46]: this function is not required.

**Review by hihigupt — APPROVED** [review]

**Pull request merged**